presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +494,53 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        // 1. normal aggregation
+        if (!preGroupedHashStrategy.isPresent()) {
+            unfinishedWork = aggregationBuilder.processPage(page);
+            return;
+        }
+
+        // 2. segmented aggregation
+        if (firstUnfinishedSegment == null) {
+            // If this is the first page, treat the first segment in this page as the current segment.
+            firstUnfinishedSegment = page.getRegion(0, 1);
+        }
+
+        Page pageOnPreGroupedChannels = page.extractChannels(preGroupedChannels);
+        int lastRowInPage = page.getPositionCount() - 1;
+        int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), pageOnPreGroupedChannels);
+        if (lastSegmentStart == 0) {
+            // The whole page is in one segment.
+            if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), 0, pageOnPreGroupedChannels)) {
+                // All rows in this page belong to the previous unfinished segment, process the whole page.
+                unfinishedWork = aggregationBuilder.processPage(page);
```

Comment:
Assume we don't have to care about the pages for a segment once it's done right? Do we need to close the `aggregationBuilder` to clear the memory after having processed a segment? It would be good to reflect the memory usage in your benchmark as well.

Reply:
> Do we need to close the aggregationBuilder to clear the memory after having processed a segment

Yes, that how it is implemented currently. Once it has fully processed at least one segments, the `aggregationBuilder` will be [closed and rebuilt](https://github.com/prestodb/presto/pull/17886/files#diff-073fe35d7e0554b1ae53998c490b8732e16d5f8f7c81d14909aa51e776c236f2R473-R478) if there are more segments to process.

> to reflect the memory usage in your benchmark as well.

Regarding the memory usage, thought the memory comparison has been covered from the manual test attached in the PR description. Could you please elaborate a bit how to run benchmark against the memory usage? Any pointers I can refer to? Thanks!